### PR TITLE
Add `-Wait` Switch parameters to the New-CaasVlan and New-CaasNetworkDomain

### PR DESCRIPTION
**Pull request by tonybaloney**

### CODE REVIEW

**Comment by tonybaloney** [comment]
cc @manishkp @BenObermayer

**Comment by tonybaloney** [comment]
hold the phone. this needs refactoring

**Comment by tonybaloney** [comment]
ok please review @manishkp 

p.s. @tintoy can you recommend a better way of having a shared "wait for thing to complete" base class/interface without having to use generics in this crazy way. This implementation feels horrible.

**Comment by tintoy** [comment]
@tonybaloney yeah, just define an interface with required properties common to all entity data-contracts. Look at the DMS powershell Cmdlets for an example of how to do this.

**Comment by tintoy** [comment]
@tonybaloney or check out [how I did it](https://github.com/DimensionDataResearch/go-dd-cloud-compute/blob/development/v1.0/compute/wait_for_status.go) in the [Go client](https://github.com/DimensionDataResearch/go-dd-cloud-compute/blob/development/v1.0/compute/resources.go).

**Comment by manishkp** [comment]
All the progress comes as public ProgressType from the API, we treat them the same, and have logic behind it.
We can add the wait logic in the base class.

**Comment by tonybaloney** [comment]
sadly `VlanType` and `NetworkDomainType` don't share a common base class, so even though they both have a `state` field, that's just a convention. Our types are all auto-documented from the XSD, so any changes I make will be overridden in the next API update.

**Comment by tonybaloney** [comment]
@tintoy @manishkp this is how I implemented it
https://github.com/tonybaloney/DimensionData.ComputeClient/blob/34641c830814b9850f2aca02ca1666ddeb47c3a8/PowershellModule/CloudCompute.Powershell/Contracts/WaitableCmdlet.cs

And an example implementation:
https://github.com/tonybaloney/DimensionData.ComputeClient/blob/34641c830814b9850f2aca02ca1666ddeb47c3a8/PowershellModule/CloudCompute.Powershell/Mcp20/DeployCaasNetworkDomainCmdlet.cs#L103-L113

**Comment by tintoy** [comment]
That's why I used an interface :)

**Comment by tonybaloney** [comment]
But you can't put the SwitchParameter field in an interface? Nor can the implementations share any kind of repeat-catch-fail logic. 
You cheated :-) 

Get Outlook for iOS




On Fri, Aug 5, 2016 at 11:14 AM +1000, "Adam Friedman" <notifications@github.com> wrote:












That's why I used an interface :)

**Comment by tintoy** [comment]
Hard-coded property on interface returns enum.

**Comment by tintoy** [comment]
@tonybaloney I've emailed you a couple of links to TFS so you can see how we did it for the Aperture powershell module (Identity or DMS are good examples).

**Comment by tonybaloney** [comment]
ok fine that solution is much better :-)

File: PowershellModule/CloudCompute.Powershell/Contracts/WaitableCmdlet.cs [hunk 1]
```diff
@@ -13,17 +15,23 @@ public abstract class WaitableCmdlet : PSCmdletCaasWithConnectionBase, IWaitable
...
                 while (!provisioned)
                 {
-                    provisioned = WaitOn<T>(objectId, ref provisionedObject);
-                    Thread.Sleep(delayTime);
+                    Update(objectId, ref provisionedObject);
+                    if (provisionedObject.state == "FAILED")
+                        ThrowTerminatingError(
+                           new ErrorRecord(new ComputeApiException(string.Format("Failed to provision {0}", provisionedObject.state)), "-1", ErrorCategory.ConnectionError, Connection));
+
+                    provisioned = (provisionedObject.state != "IN_PROGRESS" && provisionedObject.state != "PENDING_ADD");
```
Comment (manishkp) [inline PowershellModule/CloudCompute.Powershell/Contracts/WaitableCmdlet.cs]: Here is what I think our strategy should be to handle wait,
1> For "**NORMAL**" state, return **success**
2> State starts with "**PENDING_**", then we should **wait** 
3> For Rest all, we should return with **throw** if user has asked us to wait


Some of the possible states are : 
"NORMAL",
"PENDING_ADD",
"PENDING_CHANGE",
"PENDING_DELETE",
"FAILED_ADD",
"FAILED_CHANGE",
"FAILED_DELETE" and
"REQUIRES_SUPPORT".

**Comment by manishkp** [comment]
I have changed the yml file, so the build should not fail for pull requests

**Comment by tonybaloney** [comment]
@manishkp ok i rebased off develop and also implemented that logic (will kinda but the closest without having tons of nested if statemets)

**Pull request closed without merging**